File: LeetCode/752. Open the Lock/752.py
class Solution:
    def openLock(self, deadends, target):
        deadends = set(deadends)
        if "0000" in deadends:
            return -1
        
        queue = deque([("0000", 0)])
        visited = set(["0000"])
        
        while queue:
            current, turns = queue.popleft()
            if current == target:
                return turns
            
            for i in range(4):
                for j in [-1, 1]:
                    next_wheel = current[:i] + str((int(current[i]) + j) % 10) + current[i+1:]
                    if next_wheel not in visited and next_wheel not in deadends:
                        queue.append((next_wheel, turns + 1))
                        visited.add(next_wheel)
        
        return -1

File: LeetCode/752. Open the Lock/752.java
class Solution {
    public int openLock(String[] deadends, String target) {
        Set<String> deadendSet = new HashSet<>();
        for (String deadend : deadends) {
            deadendSet.add(deadend);
        }

        if (deadendSet.contains("0000")) {
            return -1;
        }

        Queue<String> queue = new LinkedList<>();
        queue.offer("0000");
        Set<String> visited = new HashSet<>();
        visited.add("0000");

        int turns = 0;

        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                String current = queue.poll();
                if (current.equals(target)) {
                    return turns;
                }

                for (int j = 0; j < 4; j++) {
                    for (int k = -1; k <= 1; k += 2) {
                        char[] currentArray = current.toCharArray();
                        int digit = (currentArray[j] - '0' + k + 10) % 10;
                        currentArray[j] = (char) (digit + '0');
                        String next = new String(currentArray);
                        if (!visited.contains(next) && !deadendSet.contains(next)) {
                            visited.add(next);
                            queue.offer(next);
                        }
                    }
                }
            }
            turns++;
        }

        return -1;
    }
}